ai/neuronal_network/dataset_preparation.py

```diff
@@ -0,0 +1,103 @@
+        if unit is not None:
+            first_player_units.append(list(unit['stat'].values()))
+    for unit in match['second_player']['units']:
+        if unit is not None:
+            second_player_units.append(list(unit['stat'].values()))
+
+    if len(first_player_units) == 0 or len(second_player_units) == 0:
+        return None
+    return {
+        'first_player_units': first_player_units,
+        'second_player_units': second_player_units,
+        'first_player_score': match['first_player']['score'],
+        'second_player_score': match['second_player']['score'],
+        'first_player_cost': match['first_player']['cost'],
+        'second_player_cost': match['second_player']['cost']
+    }
+
+
+def purge_data(data):
+    # Remove None
+    data = list(filter(lambda x: x is not None, data))
+    # Remove empty units
+    data = list(filter(lambda x: len(x['first_player_units']) != 0, data))
+    data = list(filter(lambda x: len(x['second_player_units']) != 0, data))
+    # Remove units with missing stats
+    data = list(filter(lambda x: not any(
+        len(x) != nb_stats for x in x['first_player_units']), data))
+    data = list(filter(lambda x: not any(
+        len(x) != nb_stats for x in x['second_player_units']), data))
+    # Remove matchs with score != 20
+    data = list(filter(
+        lambda x: x['first_player_score'] + x['second_player_score'] == 20, data))
+    return data
+
+
+def format_matchs(matchs):
+    max_army_len = 0
+    first_army_len = 0
+    second_army_len = 0
+    scores = []
+    units = []
+
+    for match in matchs:
+        first_army_len = len(match['first_player_units'])
+        second_army_len = len(match['second_player_units'])
+        units.append([np.array(match['first_player_units']),
+                     np.array(match['second_player_units'])])
+        scores.append([match['first_player_score'],
+                      match['second_player_score']])
+        max_army_len = max(first_army_len, second_army_len, max_army_len)
+
+    for match in units:
+        match[0] = np.pad(
+            match[0], ((0, max_army_len - len(match[0])), (0, 0)), 'constant')
+        match[1] = np.pad(
+            match[1], ((0, max_army_len - len(match[1])), (0, 0)), 'constant')
+    return (np.array(units), np.array(scores), max_army_len)
+
+
+def neuronal_network_build(shape):
+    InputModel = Input(shape=shape)
+    EncodedLayer = Dense(15, activation='relu')(InputModel)
+    EncodedLayer = Dense(15, activation='relu')(EncodedLayer)
+    EncodedLayer = Dense(15, activation='relu')(EncodedLayer)
+    EncodedLayer = Dense(15, activation='relu')(EncodedLayer)
+    EncodedLayer = Dense(15, activation='relu')(EncodedLayer)
+    EncodedLayer = Dense(15, activation='relu')(EncodedLayer)
+    EncodedLayer = Flatten()(EncodedLayer)
+    DecodedLayer = Dense(2, activation='relu')(EncodedLayer)
+    AutoEncoder = Model(InputModel, DecodedLayer)
+    AutoEncoder.compile(optimizer='adam', loss='mse')
```

Comment:
Is the optimizer a try or the more appropriate ?
Same question for the loss

Reply:
we did not conduct studies on the most best optimizer nor loss but those are the "default" ones when building a Neuronal network
like maybe not always the best ones but they do the job.
We chose them when building the first version of the ai and never move from it yet
For instance I'm currently working on an other version of the ai that works with classification and I think my loss is not the mse